plugins/sharding/logical_enumerator.rb

```diff
@@ -143,6 +187,10 @@ def key
           @key ||= Context.instance[full_key_name]
         end
 
+        def copy
+          Shard.new(@table_name, @key_name, @range_data)
+        end
```

Comment:
Umm... I need to look at this pull request deeply but I think that this code isn't a good idea...

Reply:
My another idea: In `Executor.execute`, checking `have_result_set` at the end of the `each_shard_executor` enumeration and using a shard of a current  `shard_executor`. Removing the `first_shard` variable.
The sample is below.

```ruby
    class Executor
        ...
        def execute
          enumerator = @context.enumerator

          have_result_set = false
          have_shard = false
          last_index = enumerator.size - 1
          each_shard_executor.each_with_index do |shard_executor, i|
            have_shard = true
            shard_executor.execute
            @context.consume_result_sets do |result_set|
              have_result_set = true
              yield(result_set)
            end
            next unless @context.current_limit.zero? || i >= last_index
            break if have_result_set
            result_set = HashTable.create(:flags => ObjectFlags::WITH_SUBREC,
                                          :key_type => shard_executor.shard.table)
            @context.push
            @context.temporary_tables << result_set
            targets = [[result_set]]
            @context.dynamic_columns.apply_initial(targets)
            @context.dynamic_columns.apply_filtered(targets)
            yield(result_set)
            @context.shift
          end
          ...
        end
   end
```
 
How do you feel about this approach?

Reply:
I don't like introducing `compute_size` because multiple looping implementation may be mismatched in the future.

How about simply calling `each_shard_executor` again?

```ruby
        def execute
          have_shard = false
          have_result_set = false
          each_shard_executor do |shard_executor|
            have_shard = true
            shard_executor.execute
            @context.consume_result_sets do |result_set|
              have_result_set = true
              yield(result_set)
            end
            break if @context.current_limit.zero?
          end
          unless have_shard
            enumerator = @context.enumerator
            message =
              "[logical_range_filter] no shard exists: " +
              "logical_table: <#{enumerator.logical_table}>: " +
              "shard_key: <#{enumerator.shard_key_name}>"
            raise InvalidArgument, message
          end
          return if have_result_set

          each_shard_executor do |shard_executor|
            shard = shard_executor.shard
            result_set = HashTable.create(:flags => ObjectFlags::WITH_SUBREC,
                                          :key_type => shard.table)
            @context.push
            @context.temporary_tables << result_set
            targets = [[result_set]]
            @context.dynamic_columns.apply_initial(targets)
            @context.dynamic_columns.apply_filtered(targets)
            yield(result_set)
            @context.shift
            break
          end
        end
```